proto_proxy.c

```diff
@@ -0,0 +1,3158 @@
+    if (remain <= 0) {
+        pr->key = 0;
+        pr->klen = 0;
+        return 0;
+    }
+    const char *s = memchr(cur, ' ', remain);
+    pr->key = cur - pr->request; // key offset.
+    if (s != NULL) {
+        // key is up to the next space.
+        pr->klen = s - cur;
+        if (*s == ' ') {
+            pr->has_space = true;
+        }
+    } else {
+        pr->klen = remain;
+    }
+    pr->parsed += pr->klen+1;
+
+    return 0;
+}
+
+// TODO: note TODO's from request_storage()
+static int _process_request_mset(mcp_parser_t *pr) {
+    const char *cur = pr->request + pr->parsed;
+    // see mcmc.c's _mcmc_parse_value_line() for the trick
+    // set <key> <datalen> <flags>*\r\n
+    if (!pr->has_space) {
+        return -1;
+    }
+
+    const char *s = memchr(cur, ' ', pr->reqlen - (pr->parsed + 2));
+    if (s != NULL) {
+        // Found another space, which means we at least have a key.
+        pr->key = cur - pr->request;
+        pr->klen = s - cur;
+        cur = s + 1;
+    } else {
+        return -1;
+    }
+
+    errno = 0;
+    char *n = NULL;
+    int vlen = strtol(cur, &n, 10);
+    if ((errno == ERANGE) || (cur == n)) {
+        return -1;
+    }
+    cur = n;
+
+    if (vlen < 0 || vlen > (INT_MAX - 2)) {
+       return -1;
+    }
+    vlen += 2;
+
+    pr->vlen = vlen;
+
+    return 0;
+}
+
+// TODO: error codes.
+static int _process_request_storage(mcp_parser_t *pr) {
+    const char *cur = pr->request + pr->parsed;
+    // see mcmc.c's _mcmc_parse_value_line() for the trick
+    // set <key> <flags> <exptime> <bytes> [noreply]\r\n
+    if (!pr->has_space) {
+        return -1;
+    }
+
+    // find the key. should this be done here or in main parser?
+    // here is probably better in the short term since we may end up
+    // re-parsing if ultimately passing to internal dispatch.
+    const char *s = memchr(cur, ' ', pr->reqlen - (pr->parsed + 2));
+    if (s != NULL) {
+        // Found another space, which means we at least have a key.
+        pr->key = cur - pr->request;
+        pr->klen = s - cur;
+        cur = s + 1;
+    } else {
+        return -1;
+    }
+
+    errno = 0;
+    char *n = NULL;
+    uint32_t flags = strtoul(cur, &n, 10);
+    if ((errno == ERANGE) || (cur == n) || (*n != ' ')) {
+        return -1;
+    }
+    cur = n;
+
+    errno = 0;
+    int exptime = strtol(cur, &n, 10);
+    if ((errno == ERANGE) || (cur == n) || (*n != ' ')) {
+        return -1;
+    }
+    cur = n;
+
+    errno = 0;
+    int vlen = strtol(cur, &n, 10);
+    if ((errno == ERANGE) || (cur == n)) {
+        return -1;
+    }
+    cur = n;
+
+    if (vlen < 0 || vlen > (INT_MAX - 2)) {
+       return -1;
+    }
+    vlen += 2;
+
+    // TODO: if *n is ' ' look for a CAS value.
+
+    pr->vlen = vlen;
+    pr->t.set.flags = flags;
+    pr->t.set.exptime = exptime;
+    // TODO: if next byte has a space, we check for noreply.
+    // TODO: ensure last character is \r
```

Comment:
That check might be better to put in `process_request` rather than in individual request types?

In addition to this, it may be useful to check in a proxy that

1. None of the characters before cmdlen - 2 are `\0` in the originally received command in `process_request`, which causes seems to cause issues when passed to memcache server due to memcache server expecting zero-terminated keys or commands. (If a proxy is shared by multiple clients, then misbehaving clients or bad inputs would unexpectedly affect many other clients, causing them to hang or get error/invalid responses)
2. None of the characters before cmdlen - 2 are `\r`, in case that causes issues with other proxies, future releases of memcached, or alternative servers that speak the memcached protocol.

(and return `-1` if the protocol isn't strictly followed)

Reply:
I need to triple check, but I'm pretty sure the \r\n end bits are enforced at a higher level; so down in the request parser you would look for the \r to ensure the user didn't supply junk past what the command strictly expects.